From c9aaa508757d15762bce21ea1ad76380cb3e4392 Mon Sep 17 00:00:00 2001
From: Ribesg <Ribesg@yahoo.fr>
Date: Sun, 7 Apr 2014 14:31:29 +0200
Subject: [PATCH] Prevent memory leaks.

https://github.com/SpigotMC/Spigot/blob/master/CraftBukkit-Patches/0045-Plug-World-Unload-Memory-Leak.patch
https://github.com/SpigotMC/Spigot/blob/master/CraftBukkit-Patches/0092-Reduce-memory-of-hiddenPlayers-map.patch

Modified files:
- src/main/java/net/minecraft/server/BlockRedstoneTorch.java
- src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java

diff --git a/src/main/java/net/minecraft/server/BlockRedstoneTorch.java b/src/main/java/net/minecraft/server/BlockRedstoneTorch.java
index 8e01414..e0469bb 100644
--- a/src/main/java/net/minecraft/server/BlockRedstoneTorch.java
+++ b/src/main/java/net/minecraft/server/BlockRedstoneTorch.java
@@ -11,7 +11,7 @@ import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
 public class BlockRedstoneTorch extends BlockTorch {
 
     private boolean isOn;
-    private static Map b = new HashMap();
+    private static Map b = new java.util.WeakHashMap(); // Spigot
 
     private boolean a(World world, int i, int j, int k, boolean flag) {
         if (!b.containsKey(world)) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 25aa18e..a220aed 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -64,7 +64,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     private boolean hasPlayedBefore = false;
     private final ConversationTracker conversationTracker = new ConversationTracker();
     private final Set<String> channels = new HashSet<String>();
-    private final Map<String, Player> hiddenPlayers = new MapMaker().softValues().makeMap();
+    private final Map<String, Player> hiddenPlayers = new MapMaker().weakValues().makeMap(); // Spigot - soft -> weak
     private int hash = 0;
     private double health = 20;
     private boolean scaledHealth = false;
-- 
1.8.5.2.msysgit.0
